Automatically detecting entity types in EntityBuilder

hkan opened the pull request.

phroggyy commented on src/support/Type.ts:
```diff
@@ -4,13 +4,25 @@ import { StringHelper } from './StringHelper';
 
-export function Type(type?: Function, jsonKey?: string) {
+// Types that can be passed as first argument to `EntityBuilder`.
+export type Buildable = typeof Entity | ({ new(): any });
```
```suggestion
export type Buildable = typeof Entity | ({ new<T extends Entity>(): T });
```

Can you build a non-entity – and if so, can we type it?

hkan replied:
Well, technically anything new-able can be built through EntityBuilder, so I'm not sure if it's a good idea to try to come up with a comprehensive list.

hkan commented on src/EntityBuilder.ts:
```diff
@@ -1,74 +1,69 @@
-     * @returns {any}
      */
-    public static buildOne<T>(buildClass: any, sourceData: Object): T | any {
+    public static buildOne<T extends Buildable>(buildClass: T, sourceData: object): T | object {
```
I guess `InstanceType<T> | object` as return type suffice, right? I would change async methods' return types to `InstanceType<Awaited<T>> | object`, as well.

hkan commented on src/EntityBuilder.ts:
```diff
@@ -1,74 +1,69 @@
     public static enableCamelConversion = true;
 
-    /**
+    public static buildOne<T extends Buildable>(buildClass: T, sourceData: object): T extends typeof Entity ? T : object;
```
`buildOne*` and `buildMany*` methods are used internally to resolve and fill any `@Type()`d data, as well. And there is practically no limit to what can be passed to `@Type` decorator, so it differentiates between entity classes and non-entity classes.

But I just realized there is another problem here. The return type should not be `object` if `buildClass` is non-entity. It should only be an `object` if `buildClass` is `Object`. I will fix that.

phroggyy commented on src/Entity.ts:
```diff
@@ -8,92 +7,39 @@ type CamelToSnake<T extends string> = string extends T ? string :
-        }
-        return obj;
-    }
+export type PartialEntityPropsOnly<T extends Entity> = Partial<{
```
Could you add a comment to this type, and maybe `EntityPropsOnly` too? In particular, I think what's important to highlight is that they are both recursive, and that we need `PartialEntityPropsOnly` rather than `Partial<EntityPropsOnly>` because it's recursively partial, and not only on the top level.

phroggyy commented on src/EntityBuilder.ts:
```diff
@@ -1,78 +1,88 @@
-     */
-    public static buildOne<T>(buildClass: any, sourceData: Object): T | any {
-        this.checkClassValidity(buildClass);
+    public static async buildOne<T extends Entity>(
```
I think making all these methods async reduces the library's usability. Is there a way to keep it synchronous? This kind of use greatly reduces ease-of-use as it nearly mandates use of `async` functions, and I don't think people should have to deal with promises to convert POJOs to typed classes.

hkan replied:
Asynchronousity is needed for the deferred resolution I introduced last year. I agree that it's not a great DX but having both synchronous *and* asynchronous build methods makes the codebase quite complicated with the type resolutions introduced in this MR, and it's also not a great DX either.

I'll see if I can remove asynchronous builds altogether by creating different constraints for deferred resolutions.

phroggyy commented on src/EntityBuilder.ts:
```diff
@@ -1,78 +1,88 @@
-        if (buildClass === Object) {
-            return sourceData;
-        }
+    public static async buildMany<T extends Entity>(
```
same here re async

phroggyy commented on src/support/AliasFor.ts:
```diff
@@ -1,20 +0,0 @@
-import { Entity } from '../Entity';
-
-export function AliasFor(key?: string) {
```
Why was this removed? If there's a good reason, please explain that change in the PR description, so it can be used for release notes.

hkan replied:
It's baked in to `@Type()` decorator. I thought there is very little reason to use this one other than AliasFor being more semantically reasonable.

hkan replied:
I have added extra details in the PR description.

phroggyy commented on src/support/Default.ts:
```diff
@@ -1,9 +0,0 @@
-import {defaultMetadataStorage} from "./storage";
-import {DefaultValueCallbackMetadata} from "./metadata/MetadataStorage";
-
-export function Default<T>(callback: () => T, condition: (value: T) => boolean = (value) => value === null): (target: Object, propertyKey: string) => void {
```
Same here - why is it removed?

hkan replied:
Default value is literally what you assign to the variable in the entity definition. I thought the conditioning part is probably barely used.

hkan replied:
Explained further in the PR description.

phroggyy commented on src/support/Type.ts:
```diff
@@ -4,13 +4,27 @@ import { StringHelper } from './StringHelper';
         );
 
-        const metadata = new TypeMetadata(target.constructor, key, jsonKey, type);
+        const metadata = new TypeMetadata(target.constructor as typeof Entity, key, jsonKey, type);
```
I've seen use of both `Constructor<Entity>`/`Constructor<T>` and `typeof` – is it intentional to use both? How do they differ?

phroggyy commented on src/support/isEntityType.ts:
```diff
@@ -0,0 +1,6 @@
+import { Entity } from '../Entity';
+
+export function isEntityType(buildable: Buildable): buildable is Constructor<Entity> {
+    return buildable?.prototype instanceof Entity;
```
make sure this function is tested – I know it's tiny, but it's good to have a sanity check on these simple ones too

phroggyy commented on src/support/metadata/TypeMetadata.ts:
```diff
@@ -1,26 +1,49 @@
+
+function isResolverFunction(type: Typeable): type is BuildableResolver {
+    // If the object's name is empty, we will assume it's an anonymous function that resolves the actual type.
+    return ((type as unknown as Function).name?.length === 0);
```
Why do we need to cast `as unknown as Function` when `Typeable` describes the function options?

hkan replied:
Errr, that was probably left over from earlier efforts. I tried removing it, and yeah, TS doesn't complain. Will delete that part.

phroggyy commented on tsconfig.json:
```diff
@@ -12,7 +12,7 @@
     "noUnusedLocals": true,
     "noImplicitAny": true,
-    "lib": ["es2015", "dom"],
+    "lib": ["es2015", "dom", "es2022"],
```
you can probably drop `es2015` from here if `es2022` is required.

phroggyy reviewed (changes requested):
Mostly questions and a couple small requests. My biggest objection atm is the move to *only* async support, I feel it drastically reduces the DX of the library. Happy to discuss options.

Also, as this is a major breaking change (regardless of the above), would you be able to summarise the changes so I can write release notes?

hkan commented:
I'm not particularly thrilled about the type definitions in `Entity.ts` but also couldn't come up with a better place. Any ideas?

phroggyy reviewed (commented):
I don’t see any tests for the async building, how does building with async resolution work? Can we add tests for it, or if they exist, could you point me to it?

hkan commented:
> I don’t see any tests for the async building, how does building with async resolution work? Can we add tests for it, or if they exist, could you point me to it?

Async is removed, as mentioned in this PR's description. I have also added it to changelog.md, and updated readme.md accordingly.

phroggyy commented on tsconfig.json:
```diff
@@ -12,7 +12,7 @@
     "noUnusedLocals": true,
     "noImplicitAny": true,
-    "lib": ["es2015", "dom"],
+    "lib": ["dom", "es2022"],
```
Do we need `es2022` or can we do with a less restrictive library? Would be good to keep minimal requirements where possible.

hkan replied:
Turns out we are not using much from es2022 library. Reverted existing ones to be compatible with es2015.

phroggyy reviewed (approved):
LGTM!
